File: handleMessages/src/services/messages/messages-service.js
const { Message } = require("../../db/models");

async function insertMessage(ctx, attributes) {
  const {
    source,
    messageTimestamp,
    chatType,
    chatId,
    senderId,
    isSentByMe,
    messageId,
    replyToMessageId,
    kind,
    body,
    rawSource
  } = attributes;
  //console.log(`[${ctx}] insertMessage attributes: `, { attributes });

  // Yair: don't think this makes sense long term, probably has a way to do insert-if-new in one API call.
  const existingMessage = await Message.findOne({
    where: { chatId, messageId }
  });

  if (existingMessage !== null) {
    // TODO options to update?
    return existingMessage;
  }

  const message = await Message.create({
    source,
    messageTimestamp,
    chatType,
    chatId,
    senderId,
    isSentByMe,
    messageId,
    replyToMessageId,
    kind,
    body,
    rawSource
  });

  return message;
}

async function getMessageHistory(ctx, message, options = {}) {
  const { limit = 20 } = options;
  const { chatId, messageTimestamp } = message;
  const messages = await Message.findAll({
    where: { chatId },
    limit,
    order: ["messageTimestamp"]
  });
  return messages;
}

module.exports = {
  insertMessage,
  getMessageHistory
};

File: src/services/messages/messages_service.py
from sqlalchemy import and_, desc
from src import db_models
import datetime

from src.infra.context import Context

def insert_message(ctx:Context, attributes):
    source = attributes['source']
    message_timestamp = datetime.datetime.fromtimestamp(attributes['messageTimestamp'], tz=datetime.timezone.utc)
    chat_type = attributes['chatType']
    chat_id = attributes['chatId']
    sender_id = attributes['senderId']
    is_sent_by_me = attributes['isSentByMe']
    message_id = attributes['messageId']
    reply_to_message_id = attributes['replyToMessageId']
    kind = attributes['kind']
    body = attributes['body']
    raw_source = attributes['rawSource']

    ctx.log('insertMessage attributes:', attributes)

    with db_models.Session() as session:
        existing_message = session.query(db_models.Message).filter(and_(db_models.Message.chatId == chat_id, db_models.Message.messageId == message_id)).one_or_none()

        if existing_message:
            return existing_message

        now = datetime.datetime.now()

        message = db_models.Message(
            source=source,
            messageTimestamp=message_timestamp,
            chatType=chat_type,
            chatId=chat_id,
            senderId=sender_id,
            isSentByMe=is_sent_by_me,
            messageId=message_id,
            replyToMessageId=reply_to_message_id,
            kind=kind,
            body=body,
            rawSource=raw_source,
            createdAt=now,
            updatedAt=now
        )

        session.add(message)
        session.commit()
        session.refresh(message)

        session.close()

    return message

def get_message_history(ctx:Context, message, options=None):
    if options is None:
        options = {}

    limit = options.get('limit', 20)
    chat_id = message.chatId
    message_timestamp = message.messageTimestamp

    with db_models.Session() as session:
        messages = session.query(db_models.Message) \
                   .filter(and_(db_models.Message.chatId == chat_id, db_models.Message.messageTimestamp <= message_timestamp)) \
                   .order_by(desc(db_models.Message.createdAt)).limit(limit).all()

        session.close()

    return list(reversed(messages))
